### PR TITLE
Workloadmeta-enabled Tagger Refactoring

### PR DESCRIPTION
**Pull request by juliogreff**
### What does this PR do?

After a lot of work in migrating all of the tagger to use workloadmeta, we can finally clean up the dangling bits we left around to support several collectors. There's more information in each of the individual commits, but the highlights are:

* Removing the code supporting the usage of multiple collectors themselves
* Remove parent-child relationships between entities, and ensuring each workloadmeta entity has exactly one tagger entity
* Remove tagger entity merging, made possible by the change above
* Remove sending on channels to update the tagstore, make it a simple method call, as there's only one single threaded collector now

### Additional Notes

This is a large diff, but thankfully it's mostly deletions or moving code around. Each commit is fairly self-contained and could've been a PR each, but I'm combining them all into one because there's high cohesion, and to help QA'ers lives, as the QA for each commit would've been mostly the same, and having potentially several people doing it would be wasteful. Also, if you're not Container Integrations, feel free to filter by the files you own, the changes there are small and very minor.

### Describe how to test/QA your changes

This is still a no-op from most perspectives, so from the outside the tagger still needs to work the same way it's worked before. Notable changes, or places to pay particular attention:

* `agent tagger-list` still works, but now has a single list of tags for each entity it lists. There shouldn't be any duplicated entities listed either, nor any entities with no tags. You'll also use this command to QA the rest of this PR.
* A container that belongs to a pod should have all of the pod's tags. Earlier, buggy versions of this PR sometimes missed those, and applied only the container specific k8s tags.
* A container should always have `image_*` tags. If running on a containerd k8s cluster, those tags should match the image as specified in the pod spec. It's common for containerd to resolve images to `docker.io/...` if they're not fully qualified, and those should not appear in the tags unless that's how they're specified in the pod spec. Containers that belonged to a pod, but died, no longer belong to a pod (or at least we can't tell what the pod spec was anymore), so those are fine. If the above step was ok, you can tell them apart by the lack of k8s tags.
* A container that belongs to an ECS task should have ECS tags applied. They're the `task_*` ones, among others.
* Telemetry works. `datadog.agent.tagger_entities` should match the number of tags in the store. This should be checked after enough workloads lived and died to make sure there's no drift.
* Entity pruning works. After a container has been deleted (actually deleted, containerd keeps them around after they've died for example), its entity in the tagger should be gone about 5 minutes afterwards. After a pod is gone (and presumably its containers have been deleted as well), check that the pod's containers are also gone from the tagger.
* Check that there are no duplicated tags. It would be obvious from looking at the code that this is not possible, but it never hurts to double check.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: cmd/agent/api/response/types.go [hunk 1]
```diff
@@ -24,5 +24,5 @@ type TaggerListResponse struct {
 
 // TaggerListEntity holds the tagging info about an entity
 type TaggerListEntity struct {
-	Tags map[string][]string `json:"tags"`
```
Comment (djmitche) [inline cmd/agent/api/response/types.go:27]: I see this has a `json` annotation -- is this changing the format of any external JSON?

**Review by djmitche — APPROVED** [review]
agent-core files look good

File: cmd/agent/api/response/types.go (hunk 1 above)
Comment (juliogreff) [reply cmd/agent/api/response/types.go:27]: Yes, this is part of the flare payload. I'm not aware of any automated consumers (and this changed not too far back anyway, to no noise whatsoever), so I'm assuming it's only read by humans, and most likely humans working at Datadog.

File: cmd/agent/app/tagger_list.go [hunk 2]
```diff
@@ -75,38 +75,24 @@ var taggerListCommand = &cobra.Command{
 		for entity, tagItem := range tr.Entities {
 			fmt.Fprintln(color.Output, fmt.Sprintf("\n=== Entity %s ===", color.GreenString(entity)))
 
-			sources := make([]string, 0, len(tagItem.Tags))
-			for source := range tagItem.Tags {
-				sources = append(sources, source)
-			}
+			fmt.Fprint(color.Output, "Tags: [")
```
Comment (clamoriniere) [inline cmd/agent/app/tagger_list.go:78]: nit: maybe we can use a `strings.Builder` here.

File: cmd/agent/common/loader.go [hunk 3]
```diff
@@ -28,12 +27,11 @@ import (
...
 	if flavor.GetFlavor() != flavor.ClusterAgent {
-		workloadmeta.GetGlobalStore().Start(ctx)
+		store := workloadmeta.GetGlobalStore()
+		store.Start(ctx)
 
-		// start the tagger. must be done before autodiscovery, as it needs to
-		// be the first subscribed to metadata store to avoid race conditions.
-		tagger.SetDefaultTagger(local.NewTagger(collectors.DefaultCatalog))
-		if err := tagger.Init(); err != nil {
+		tagger.SetDefaultTagger(local.NewTagger(store))
```
Comment (clamoriniere) [inline cmd/agent/common/loader.go:33]: Should we keep the previous comment about the subscriber ordering?

File: cmd/dogstatsd/main.go [hunk 4]
```diff
@@ -213,11 +212,11 @@ func runAgent(ctx context.Context) (err error) {
 
 	// container tagging initialisation if origin detection is on
 	if config.Datadog.GetBool("dogstatsd_origin_detection") {
-		// Start workload metadata store before tagger
-		workloadmeta.GetGlobalStore().Start(context.Background())
+		store := workloadmeta.GetGlobalStore()
```
Comment (clamoriniere) [inline cmd/dogstatsd/main.go:215]: same here about previous comment.

File: pkg/logs/internal/tailers/journald/docker.go [hunk 5]
```diff
@@ -43,7 +45,7 @@ func (t *Tailer) getContainerTags(containerID string) []string {
 
 // initializeTagger initializes the tag collector.
 func (t *Tailer) initializeTagger() {
-	err := tagger.Init()
+	err := tagger.Init(context.TODO())
```
Comment (clamoriniere) [inline pkg/logs/internal/tailers/journald/docker.go:48]: should we update `initializeTagger()` to provide a context?

Comment (clamoriniere) [reply pkg/logs/internal/tailers/journald/docker.go:48]: Ok I checked https://github.com/DataDog/datadog-agent/blob/55da4f3c2a2dfce0f7e57c1fcb5c590636c53a88/pkg/logs/internal/tailers/journald/tailer.go#L76 
and it seems to be a big change to make it properly.

File: pkg/tagger/collectors/workloadmeta_extract.go [hunk 6]
```diff
@@ -102,82 +105,94 @@ var (
...
+		}
 	}
 
-	// NOTE: haha, this is still async and race conditions will still
-	// happen :D since the workloadmeta will be the only collector in the
-	// tagger in the end, this can be turned into a sync call to
-	// processTagInfo
 	if len(tagInfos) > 0 {
-		c.out <- tagInfos
+		c.p.ProcessTagInfo(tagInfos)
```
Comment (clamoriniere) [inline pkg/tagger/collectors/workloadmeta_extract.go:144]: with `c.p` is not clear what is `p`, can we rename it with a more self explain name?

File: pkg/tagger/collectors/workloadmeta_main.go [hunk 7]
```diff
@@ -20,22 +20,15 @@ import (
...
 
 // WorkloadMetaCollector collects tags from the metadata in the workloadmeta
 // store.
 type WorkloadMetaCollector struct {
-	store    workloadmeta.Store
-	children map[string]map[string]struct{}
-	out      chan<- []*TagInfo
-	stop     chan struct{}
+	store workloadmeta.Store
+	p     processor
```
Comment (clamoriniere) [inline pkg/tagger/collectors/workloadmeta_main.go:31]: maybe worth to use the full name.

```suggestion
	processor     processor
```

**Review by clamoriniere — COMMENTED** [review]
Cleaner and clearer code, I like that 💯 

only few nits

**Comment by juliogreff** [comment]
Closed in favor of #11309

**Pull request closed without merging**